File: Data Structures & Algorithms/house-robber-ii/submission-4.py
class Solution:
    def rob(self, nums: List[int]) -> int:

        def helper(nums: List[int]) -> int:

            n = len(nums)
            dp = [0] * (n+2)

            for i in range(n-1, -1, -1):
                dp[i] = max(nums[i] + dp[i+2], dp[i+1])

            return dp[0]


        if len(nums) == 1: return nums[0]
        return max(helper(nums[1:]), helper(nums[:-1]))
        

File: Data Structures & Algorithms/longest-consecutive-sequence/submission-4.java
public class Solution {
    public int longestConsecutive(int[] nums) {
        
        Set<Integer> numSet = new HashSet<>();
        for (int num : nums) numSet.add(num);
        
        int longest = 0;
        for (int num : numSet) {
            if (!numSet.contains(num - 1)) {
                int length = 1;
                while (numSet.contains(num + length)) {
                    length++;
                }
                longest = Math.max(longest, length);
            }
        }
        return longest;
    }
}



// class Solution {
//     public int longestConsecutive(int[] nums) {

//         if (nums.length == 0) return 0;

//         // O(nlogn) time complexity due to sorting
//         Arrays.sort(nums);

//         int maxSeq = 1; // The longest sequence found
//         int seq = 1; // Current sequence length

//         for (int i = 1; i < nums.length; i++) {

//             // Skip duplicates
//             if (nums[i] == nums[i - 1]) {
//                 continue;
//             }

//             // If the current number is consecutive to the previous one
//             if (nums[i] == nums[i - 1] + 1) {
//                 seq++;
//             } else {
//                 // Reset the sequence length if it's not consecutive
//                 maxSeq = Math.max(maxSeq, seq);
//                 seq = 1; // Start a new sequence
//             }
//         }

//         // Final check for the last sequence
//         return Math.max(maxSeq, seq);
//     }
// }
